src/nvim/plines.c

```diff
@@ -991,16 +994,17 @@ int64_t win_text_height(win_T *const wp, const linenr_T start_lnum, const int64_
     lnum = lnum_next + 1;
   }
 
-  while (lnum <= end_lnum) {
+  int64_t max_height = max != NULL ? *max : INT64_MAX;
+  while (lnum <= end_lnum && height_sum_nofill + height_sum_fill < max_height) {
     linenr_T lnum_next = lnum;
     hasFolding(wp, lnum, &lnum, &lnum_next);
     height_sum_fill += win_get_fill(wp, lnum);
     height_cur_nofill = plines_win_nofill(wp, lnum, false);
     height_sum_nofill += height_cur_nofill;
+    lnum_prev = lnum;
     lnum = lnum_next + 1;
   }
-
-  if (end_vcol >= 0) {
+  if (end_vcol >= 0 && lnum_prev == end_lnum) {
```

Comment:
This does not work if `end_lnum` is inside a multi-line fold, as `lnum_prev` is at the start of the fold.